pkg/virt-controller/watch/migration.go

```diff
@@ -0,0 +1,717 @@
+	}
+
+	c.vmiInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addVMI,
+		DeleteFunc: c.deleteVMI,
+		UpdateFunc: c.updateVMI,
+	})
+
+	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addPod,
+		DeleteFunc: c.deletePod,
+		UpdateFunc: c.updatePod,
+	})
+
+	c.migrationInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addMigration,
+		DeleteFunc: c.deleteMigration,
+		UpdateFunc: c.updateMigration,
+	})
+
+	return c
+}
+
+func (c *MigrationController) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.Queue.ShutDown()
+	log.Log.Info("Starting migration controller.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmiInformer.HasSynced, c.podInformer.HasSynced, c.migrationInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping migration controller.")
+}
+
+func (c *MigrationController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *MigrationController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing Migration %v", key)
+		c.Queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed Migration %v", key)
+		c.Queue.Forget(key)
+	}
+	return true
+}
+
+func (c *MigrationController) execute(key string) error {
+	var vmi *virtv1.VirtualMachineInstance
+	var targetPods []*k8sv1.Pod
+
+	// Fetch the latest state from cache
+	obj, exists, err := c.migrationInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		c.podExpectations.DeleteExpectations(key)
+		return nil
+	}
+	migration := obj.(*virtv1.VirtualMachineInstanceMigration)
+	logger := log.Log.Object(migration)
+
+	vmiObj, vmiExists, err := c.vmiInformer.GetStore().GetByKey(fmt.Sprintf("%s/%s", migration.Namespace, migration.Spec.VMIName))
+	if err != nil {
+		return err
+	}
+
+	if vmiExists {
+		vmi = vmiObj.(*virtv1.VirtualMachineInstance)
+		targetPods, err = c.listMatchingTargetPods(migration, vmi)
+		if err != nil {
+			return err
+		}
+	}
+
+	needsSync := c.podExpectations.SatisfiedExpectations(key) && vmiExists
+
+	logger.V(4).Infof("processing migration: needsSync %t, hasVMI %t, targetPod len %d", needsSync, vmiExists, len(targetPods))
+
+	var syncErr error
+
+	if needsSync && !migration.IsFinal() {
+		syncErr = c.sync(migration, vmi, targetPods)
+	}
+
+	err = c.updateStatus(migration, vmi, targetPods, syncErr)
+	if err != nil {
+		return err
+	}
+
+	if syncErr != nil {
+		return syncErr
+	}
+
+	return nil
+}
+
+func (c *MigrationController) canMigrateVMI(migration *virtv1.VirtualMachineInstanceMigration, vmi *virtv1.VirtualMachineInstance) (bool, error) {
+
+	if vmi.Status.MigrationState == nil {
```

Comment:
I didn't find any checks if the VMI is running. Are we leaving all that to virt-handlers? If I didn't miss some checks, wouldn't we create target pods for `Failed` and `Succeeded` VMIs too?

Reply:
This function was meant to ensure that no other migrations were still in progress. There's logic in the function that looks at updating the Migration's status that fails the migration job immediately if the VMI moves to a final state. The migration object itself can't even be introduced into the cluster unless the VMI is in the running state.